feat: Add flipt provider with PSR-16 caching support

legoheld opened the pull request:
## This PR
Adds a provider for https://www.flipt.io/ with PSR-16 caching support. 

### How to test
How to run tests is documented in the README.md

tcarrio committed:
docs: update README.md to include integrations terminology

Signed-off-by: Tom Carrio <tom@carrio.dev>
Signed-off-by: legoheld <dreckig@sandkasten.ch>

legoheld committed:
feat: Add flipt provider with PSR-16 caching support

Signed-off-by: legoheld <dreckig@sandkasten.ch>

legoheld committed:
feat: Add flipt provider with PSR-16 caching support

Signed-off-by: legoheld <dreckig@sandkasten.ch>

markphelps commented on .vscode/launch.json:
```diff
@@ -0,0 +1,51 @@
+{
+    // Use IntelliSense to learn about possible attributes.
```
may want to remove this file/folder

markphelps commented on providers/Flipt/LICENSE:
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
mind updating this to MIT? its what we use for most of our client sdks

beeme1mr replied:
Apache 2 is the recommended license for cncf projects.

markphelps replied:
Thanks @beeme1mr sounds good

markphelps reviewed (commented):
minor things, looks good otherwise from a Flipt perspective! thanks @legoheld !!

legoheld committed:
docs: Change to MIT license

Signed-off-by: legoheld <dreckig@sandkasten.ch>

legoheld committed:
docs: Change to MIT license

Signed-off-by: legoheld <dreckig@sandkasten.ch>

legoheld committed:
fix: Remove unnecessary launch.json

Signed-off-by: legoheld <dreckig@sandkasten.ch>

legoheld committed:
fix: Remove unnecessary launch.json

Signed-off-by: legoheld <dreckig@sandkasten.ch>

legoheld committed:
docs: Revert back to apache2 license according to @beeme1mr

Signed-off-by: legoheld <dreckig@sandkasten.ch>

legoheld committed:
docs: Revert back to apache2 license according to @beeme1mr

Signed-off-by: legoheld <dreckig@sandkasten.ch>

tcarrio commented on providers/Flipt/src/Cache.php:
```diff
@@ -0,0 +1,64 @@
+
+
+    public function key( array $params ) {
+        return md5( json_encode( $params ) );
```
Just making a note that this is not collision free.

This could have unintended side effects in a production system with many evaluations that might not assume caching with Flipt would run into that issue.

tcarrio replied:
Also, without normalizing the object you are open to cache misses for equivalent objects.

Rudimentary example of this here: https://onlinephp.io/c/04641

You would normalize it by consistently mapping the params to the same output string. This might require some canonicalized sorting of keys beforehand.

legoheld replied:
Very good hint here.

If added a recursive ksort on the context attributes, so the cache misses should be avoided.

tcarrio commented on providers/Flipt/src/FliptProvider.php:
```diff
@@ -0,0 +1,137 @@
+    protected $client;
+    protected Cache $cache;
+
+    public function __construct( mixed $hostOrClient, string $apiToken = '', string $namespace = '', CacheInterface $cache = null ) {
```
The minimum supported version of OpenFeature SDK is PHP 8.0, so you can used union types here.


```suggestion
    public function __construct(string|FliptClient $hostOrClient, string $apiToken = '', string $namespace = '', CacheInterface $cache = null ) {
```

tcarrio commented on providers/Flipt/src/FliptProvider.php:
```diff
@@ -0,0 +1,137 @@
+        
+        // there is a match
+        // not sure yet as the variant result has a getMatch() but not the boolean result.
+        if( $result->getReason() == 'MATCH_EVALUATION_REASON' || $result->getReason() == "DEFAULT_EVALUATION_REASON" ) {
```
Could we set these strings up as consts, if they aren't available from the Flipt SDK directly?

tcarrio commented on providers/Flipt/src/FliptProvider.php:
```diff
@@ -0,0 +1,137 @@
+                return json_decode( $result->getVariantAttachment(), true);
+            case FlagValueType::STRING:
+                return $result->getVariantKey();
+            default:
```
This should never happen, in which case if it did, it should be an error not nullng the value.

legoheld replied:
Im not sure if this is an issue as I use this only when I get a successfull response from the client:
```php
if( $result->getReason() == ResponseReasons::MATCH_EVALUATION_REASON || $result->getReason() == ResponseReasons::DEFAULT_EVALUATION_REASON ) {
            $result = ResolutionDetailsFactory::fromSuccess( $this->castResult( $result, $flagType ) );
}
```

And the I suppose a correct value. But you are in a way right if the response value is somewhat invalid I should throw an error. What do you propose here?

tcarrio commented on providers/Flipt/src/Cache.php:
```diff
@@ -0,0 +1,64 @@
+
+        if( empty( $this->cache ) ) return;
+
+        $entries = $this->cache->get( self::CACHE_KEY, []);
+        $entries[ $key ] = $value;
+
+        $this->cache->set( self::CACHE_KEY, $entries );
```
This introduces race conditions for distributed storage caches like Redis or Memcached. You can atomically interact with the cache per key if you utilize a key prefix instead.


```suggestion
        $this->cache->set($key, $value);
```

This does not fix the issue I mentioned below about the collisions. To resolve both, you would still need a different solution.

tcarrio replied:
Or, utilize a sufficiently large range hash function to avoid collisions. This might be something to consider being a configurable setting, e.g. raw JSON encoded keys or md5 or sha256 would all be valid strategies, to give you the overall safety necessary for the application.

legoheld replied:
Now this is the hard part. With PSR-16 there is no way to delete all keys with a certain cache key prefix.
So I have to ways to solve that either the way you have seen ( all entries under one key ) or I store the keys to delete under a separate cache key. But also there you have racing conditions as you mentioned. But the later is probably more efficient.

I actually dont know how to solve that with PSR-16. Maybe just by documenting that this is not ready to use with a distributed caching storage.

legoheld replied:
Dedicated cahce storage -> clear will clear all keys.

tcarrio commented on providers/Flipt/tests/FliptProviderTest.php:
```diff
@@ -0,0 +1,147 @@
+    }
+
+
+    public function testCache()
```
I would suggest having test files per class. The cache being separate, it should exist in a `CacheTest.php` separate from the provider tests

legoheld replied:
I refactored the whole Caching mechanism into its own CacheProvider.
This is now a wrapper around other Providers to allow caching. This way the FliptProvider does not know anything about caching und could be an approach for https://github.com/open-feature/php-sdk/issues/106.

I also remove the tests in it own file.

tcarrio commented on providers/Flipt/README.md:
```diff
@@ -0,0 +1,60 @@
+OpenFeature::setProvider($provider);
+
+// Now you can evaluate your feature flags as follows
+$booleanFlagValue = OpenFeature::getBooleanValue('your-boolean-flag-key', false);
+$stringFlagValue = OpenFeature::getStringValue('your-string-flag-key', 'default-value');
+$integerFlagValue = OpenFeature::getIntegerValue('your-integer-flag-key', 0);
+$floatFlagValue = OpenFeature::getFloatValue('your-float-flag-key', 0.0);
+$objectFlagValue = OpenFeature::getObjectValue('your-object-flag-key', ['default' => 'value']);
```
You need to retrieve a client currently, these resolve methods are not on the global API.

tcarrio replied:
Instead of `OpenFeature::getBoolean` you would do `$client = OpenFeature::getInstance()->get client()` and interact with that.

tcarrio replied:
A facade global might be a useful functionality but it's not an available one right now 🙂

legoheld replied:
Actually I copied the wrong code. Thats the usage when using my [laravel openfeature package](https://github.com/legoheld/openfeature-laravel) ;)

There I registered the OpenFeature client directly under a facade for easy access.

tcarrio reviewed (commented):
Thanks for the pull request! It's looking pretty good so far, mostly have questions around the caching provided thus far.

Commit:
feat: add CacheProvider to extract caching funcionality from FliptProvider
fix: Add ResponseReasons constants
fix: proper FliptProvider constructor type
fix: Add ksort for context attributes

Commit:
feat: add CacheProvider to extract caching funcionality from FliptProvider
fix: Add ResponseReasons constants
fix: proper FliptProvider constructor type
fix: Add ksort for context attributes

Signed-off-by: legoheld <dreckig@sandkasten>

Commit:
feat: add CacheProvider to extract caching funcionality from FliptProvider
fix: Add ResponseReasons constants
fix: proper FliptProvider constructor type
fix: Add ksort for context attributes

Signed-off-by: legoheld <dreckig@sandkasten>
Signed-off-by: legoheld <dreckig@sandkasten.ch>

Commit:
Merge branch 'main' of https://github.com/legoheld/php-sdk-contrib

Commit:
chore: fix readme

Signed-off-by: legoheld <dreckig@sandkasten>

Commit:
chore: fix readme

Signed-off-by: legoheld <dreckig@sandkasten>
Signed-off-by: legoheld <dreckig@sandkasten.ch>

legoheld commented:
Thank you @tcarrio for this very detailed feedback!

markphelps commented on providers/Flipt/composer.json:
```diff
@@ -0,0 +1,34 @@
+        "open-feature",
+        "feature flags"
+    ],
+    "homepage": "https://github.com/legoheld/flipt-provider",
```
should this be this repo? or our (Flipt's) website?

legoheld replied:
Remove the homepage as its optional. We do it like the other providers.

markphelps commented on providers/Flipt/composer.json:
```diff
@@ -0,0 +1,34 @@
+    "require": {
+        "php": ">=8.0",
+        "open-feature/sdk": "^2.0",
+        "flipt-io/flipt": "^0.0.1",
```
just released `0.2.0` btw https://github.com/flipt-io/flipt-php/releases/tag/v0.2.0

legoheld replied:
@markphelps Updated package to use ResponseConstants from client.

PS: Sorry about the missing conventional commit ;)

Commit:
chore: Update to client lib constants

Signed-off-by: legoheld <dreckig@sandkasten>

Commit:
chore: Update to client lib constants

Signed-off-by: legoheld <dreckig@sandkasten>
Signed-off-by: legoheld <dreckig@sandkasten.ch>

legoheld committed:
Merge branch 'main' of https://github.com/legoheld/php-sdk-contrib

legoheld committed:
fix: proper response reasons in test

Signed-off-by: legoheld <dreckig@sandkasten.ch>

legoheld committed:
doc: adjust caching infos according to new caching provider approach

Signed-off-by: legoheld <dreckig@sandkasten.ch>

tcarrio commented:
A few things I noticed after we talked @legoheld :

- unit tests exist but do not run in CI. Include the Flipt provider package in the matrix for testing in GH Actions
- I will need to create the corresponding repository this will "deploy" through before we merge. We gitsplit the packages to individual repos hosted under @open-feauture-php

tcarrio commented:
The [new repository](https://github.com/open-feature-php/flipt-provider) has been created.

You will need to point the gitsplit config for the new Flipt provider to it in the `.gitsplit.yml` file

markphelps commented:
👋🏻 Is there anything I can do to help get this across the line? much appreciated @legoheld and @tcarrio !!

tcarrio commented:
@legoheld Just reaching out to see if this is something you still wanted to continue with. If so, then the commits would need to be assigned in order to contribute the work.

Let me know! 👋

legoheld commented:
Thanks for the reminder @tcarrio.

Im quite busy with projects and havent caught up with the flip progress.
@markphelps would you like to take over?

markphelps commented:
> Thanks for the reminder @tcarrio.
> 
> Im quite busy with projects and havent caught up with the flip progress. @markphelps would you like to take over?

Yes I will take over. Thank you @legoheld !!

tcarrio commented:
No worries, that works fine @legoheld. If the work is being continued from your branch then you should sign off the commits for the DCO.

If you have questions @markphelps throw them in the thread and I'll get to it as soon as I can 👋
